src/platform/riot/uart_channel.c

```diff
@@ -0,0 +1,163 @@
+                                                                               const FederateMessage *msg),
+                                                      FederatedConnectionBundle *conn) {
+  UART_CHANNEL_INFO("Register receive callback");
+  UARTPollChannel *self = (UARTPollChannel *)untyped_self;
+
+  self->receive_callback = receive_callback;
+  self->federated_connection = conn;
+}
+
+void _UARTPollChannel_interrupt_callback(void *arg, uint8_t received_byte) {
+  UARTPollChannel *self = (UARTPollChannel *)arg;
+  const uint32_t minimum_message_size = 12;
+
+  self->receive_buffer[self->receive_buffer_index] = received_byte;
+  self->receive_buffer_index++;
+
+  if (self->receive_buffer_index >= minimum_message_size) {
+    if (self->super.super.mode == NETWORK_CHANNEL_MODE_ASYNC) {
+      cond_signal(&((UARTAsyncChannel *)self)->receive_cv);
+    }
+  }
+}
+
+void UARTPollChannel_poll(NetworkChannel *untyped_self) {
+  UARTPollChannel *self = (UARTPollChannel *)untyped_self;
+  const uint32_t minimum_message_size = 12;
+
+  while (self->receive_buffer_index > minimum_message_size) {
+    int bytes_left = deserialize_from_protobuf(&self->output, self->receive_buffer, self->receive_buffer_index);
+    UART_CHANNEL_DEBUG("Bytes Left after attempted to deserialize %d", bytes_left);
+
+    if (bytes_left >= 0) {
+      int receive_buffer_index = self->receive_buffer_index;
+      self->receive_buffer_index = bytes_left;
+      memcpy(self->receive_buffer, self->receive_buffer + (receive_buffer_index - bytes_left), bytes_left);
+
+      if (self->receive_callback != NULL) {
+        UART_CHANNEL_DEBUG("calling user callback!");
+        self->receive_callback(self->federated_connection, &self->output);
+      }
+    }
+  }
+}
+
+void *_UARTAsyncChannel_decode_loop(void *arg) {
+  UARTAsyncChannel *self = (UARTAsyncChannel *)arg;
+  mutex_lock(&self->receive_lock);
+
+  UART_CHANNEL_DEBUG("Entering decoding loop");
+  while (true) {
+    cond_wait(&self->receive_cv, &self->receive_lock);
+
+    UARTPollChannel_poll((NetworkChannel *)&self->super);
+  }
+
+  return NULL;
+}
+
+void UARTPollChannel_ctor(UARTPollChannel *self, Environment *env, uint32_t uart_device, uint32_t baud) {
+  assert(self != NULL);
+  assert(env != NULL);
+
+  self->uart_dev = UART_DEV(uart_device);
+
+  int result = uart_init(self->uart_dev, baud, _UARTPollChannel_interrupt_callback, self);
+
+  if (result == -ENODEV) {
+    UART_CHANNEL_ERR("Invalid UART device!");
+    throw("The user specified an invalid UART device!");
+  } else if (result == ENOTSUP) {
+    UART_CHANNEL_ERR("Given configuration to uart is not supported!");
+    throw("The given combination of parameters for creating a uart device is unsupported!");
+  } else if (result < 0) {
+    UART_CHANNEL_ERR("UART Init error occurred");
+    throw("Unknown UART RIOT Error!");
+  }
+
+  result = uart_mode(self->uart_dev, UART_DATA_BITS_8, UART_PARITY_EVEN, UART_STOP_BITS_2);
```

Comment:
These options should be possible to set from the lf file, I think something like:

```
@interface_uart(device_id=1, baudrate=115200, data_bits=8, parity="none"/"even"/"odd", stop_bits=1)
f = new UartFed()
```

I.e. they must be exposed through the UartPollChannel_ctor. But we shouldnt expose the RIOT specific types, because we want to provide implementations for Zephyr and Posix also. So you might need to create your own typedefs for this.

Also, the normal default is 8N1 (8 data bits, no parity bit, and 1 stop bit)

Reply:
done, there are now integer parameters to the constructors.

Reply:
done